heed/src/database.rs

```diff
@@ -1993,6 +1996,265 @@ impl<KC, DC, C> Database<KC, DC, C> {
+    ///
+    /// The entry is written with the specified flags.
+    ///
+    /// ```
+    /// # use heed::EnvOpenOptions;
+    /// use std::io::Write;
+    /// use heed::{Database, PutFlags};
+    /// use heed::types::*;
+    /// use heed::byteorder::BigEndian;
+    ///
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// # let dir = tempfile::tempdir()?;
+    /// # let env = unsafe { EnvOpenOptions::new()
+    /// #     .map_size(10 * 1024 * 1024) // 10MB
+    /// #     .max_dbs(3000)
+    /// #     .open(dir.path())?
+    /// # };
+    /// type BEI32 = I32<BigEndian>;
+    ///
+    /// let mut wtxn = env.write_txn()?;
+    /// let db = env.create_database::<BEI32, Str>(&mut wtxn, Some("number-string"))?;
+    ///
+    /// # db.clear(&mut wtxn)?;
+    /// let long = "I am a long long long value";
+    /// assert_eq!(
+    ///     db.get_or_put_reserved_with_flags(&mut wtxn, PutFlags::empty(), &42, long.len(), |reserved| {
+    ///         reserved.write_all(long.as_bytes())
+    ///     })?,
+    ///     None
+    /// );
+    ///
+    /// let longer = "I am an even longer long long long value";
+    /// assert_eq!(
+    ///     db.get_or_put_reserved_with_flags(&mut wtxn, PutFlags::empty(), &42, longer.len(), |reserved| {
+    ///         unreachable!()
+    ///     })?,
+    ///     Some(long)
+    /// );
+    ///
+    /// let ret = db.get(&mut wtxn, &42)?;
+    /// assert_eq!(ret, Some(long));
+    ///
+    /// wtxn.commit()?;
+    /// # Ok(()) }
+    /// ```
+    pub fn get_or_put_reserved_with_flags<'a, 'txn, F>(
+        &'txn self,
+        txn: &mut RwTxn,
+        flags: PutFlags,
+        key: &'a KC::EItem,
+        data_size: usize,
+        write_func: F,
+    ) -> Result<Option<DC::DItem>>
+    where
+        KC: BytesEncode<'a>,
+        F: FnOnce(&mut ReservedSpace) -> io::Result<()>,
+        DC: BytesDecode<'a>,
+    {
+        assert_eq_env_db_txn!(self, txn);
+
+        let key_bytes: Cow<[u8]> = KC::bytes_encode(key).map_err(Error::Encoding)?;
+
+        let mut key_val = unsafe { crate::into_val(&key_bytes) };
+        let mut reserved = ffi::reserve_size_val(data_size);
+        let flags = (flags | PutFlags::NO_OVERWRITE).bits() | lmdb_master_sys::MDB_RESERVE;
+
+        unsafe {
+            match ffi::mdb_put(txn.txn.txn, self.dbi, &mut key_val, &mut reserved, flags) {
+                lmdb_master_sys::MDB_KEYEXIST => {
+                    let bytes = crate::from_val(reserved);
+                    let data = DC::bytes_decode(bytes).map_err(Error::Decoding)?;
+                    return Ok(Some(data));
+                }
+                err_code => mdb_result(err_code)?,
+            }
+        }
+
+        let mut reserved = unsafe { ReservedSpace::from_val(reserved) };
+        write_func(&mut reserved)?;
+        if reserved.remaining() == 0 {
+            Ok(None)
+        } else {
+            Err(io::Error::from(io::ErrorKind::UnexpectedEof).into())
```

Comment:
We could also consider adding a new variant to the [`Error`](https://github.com/meilisearch/heed/blob/7e2e105faef1795f2ec7a42bdc0724b1e95146f8/heed/src/lib.rs#L143) enum. Something like (but maybe with a better name):
```
Error::ReservedShortWrite {
    written: usize,
    reserved: usize
}
```
That way we don't have to allocate a new `io::Error`, but still get a helpful message. 

The reason we should avoid allocations is that short writes could be a valid use-case and may even be in a hot-path:
- A user may reserve more space than they'll need, and then fill the buffer with some data, keeping track themselves of how much data has actually been written (either by a header or sentinel-terminator).

However, reading through the docs for `MDB_RESERVE`, we may want to write zeroes to all unwritten bytes instead of leaving them untouched:
> the caller is expected to modify all of the space requested

Reply:
Relates issue https://github.com/meilisearch/heed/issues/254.